File: ExoPlay2/src/main/java/com/ashlikun/media/exoplay2/play/EasyVideoExo2.java
package com.ashlikun.media.exoplay2.play;

import android.media.AudioManager;
import android.net.Uri;
import android.text.TextUtils;
import android.view.Surface;
import android.widget.Toast;

import com.ashlikun.media.R;
import com.ashlikun.media.exoplay2.IjkExo2MediaPlayer;
import com.ashlikun.media.video.EasyMediaInterface;
import com.ashlikun.media.video.EasyMediaManager;

import java.io.File;

import tv.danmaku.ijk.media.player.IMediaPlayer;

/**
 * 作者　　: 李坤
 * 创建时间: 2017/12/01 16:25
 * 邮箱　　：496546144@qq.com
 * <p>
 * 功能介绍：实现Exo2的播放引擎
 */
public class EasyVideoExo2 extends EasyMediaInterface
        implements IjkExo2MediaPlayer.OnPreparedListener,
        IjkExo2MediaPlayer.OnCompletionListener,
        IjkExo2MediaPlayer.OnBufferingUpdateListener,
        IjkExo2MediaPlayer.OnErrorListener,
        IjkExo2MediaPlayer.OnInfoListener,
        IjkExo2MediaPlayer.OnSeekCompleteListener,
        IjkExo2MediaPlayer.OnVideoSizeChangedListener {
    /**
     * 是否调用过暂定，调用后在准备好的时候不能直接播放
     */
    private boolean isPreparedPause = false;
    public IjkExo2MediaPlayer mediaPlayer;

    @Override
    public void setPreparedPause(boolean preparedPause) {
        isPreparedPause = preparedPause;
    }

    @Override
    public void start() {
        //暂停不在同一个管理器的播放器
        EasyMediaManager.pauseOther(easyMediaManager);
        if (mediaPlayer != null) {
            mediaPlayer.start();
        }
    }

    @Override
    public void stop() {
        if (mediaPlayer != null) {
            mediaPlayer.stop();
        }
    }

    @Override
    public void prepare() {
        try {
            if (isPlaying()) {
                stop();
            }
            release();
        } catch (Exception e) {
        }
        mediaPlayer = new IjkExo2MediaPlayer(context);

        if (getCurrentDataSource() == null) {
            Toast.makeText(context, context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show();
            onError(mediaPlayer, -2, -2);
            return;
        }

        mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
        mediaPlayer.setOnPreparedListener(EasyVideoExo2.this);
        mediaPlayer.setOnCompletionListener(EasyVideoExo2.this);
        mediaPlayer.setOnBufferingUpdateListener(EasyVideoExo2.this);
        mediaPlayer.setScreenOnWhilePlaying(true);
        mediaPlayer.setOnErrorListener(EasyVideoExo2.this);
        mediaPlayer.setOnInfoListener(EasyVideoExo2.this);
        mediaPlayer.setOnVideoSizeChangedListener(EasyVideoExo2.this);
        mediaPlayer.setOnSeekCompleteListener(this);

        mediaPlayer.setLooping(getCurrentDataSource().isLooping());

        //通过自己的内部缓存机制
        mediaPlayer.setCache(getCurrentDataSource().isCache());
        if (!TextUtils.isEmpty(getCurrentDataSource().cacheDir())) {
            mediaPlayer.setCacheDir(new File(getCurrentDataSource().cacheDir()));
        }
        mediaPlayer.setOverrideExtension(getCurrentDataSource().overrideExtension());

        try {
            if (!TextUtils.isEmpty(getCurrentDataSource().getUrl())) {
                if (getCurrentDataSource().getHeaders() != null) {
                    mediaPlayer.setDataSource(context, Uri.parse(getCurrentDataSource().getUrl()), getCurrentDataSource().getHeaders());
                } else {
                    mediaPlayer.setDataSource(getCurrentDataSource().getUrl());
                }
            } else if (getCurrentDataSource().getUri() != null && !TextUtils.isEmpty(getCurrentDataSource().getUri().toString())) {
                if (getCurrentDataSource().getHeaders() != null) {
                    mediaPlayer.setDataSource(context, getCurrentDataSource().getUri(), getCurrentDataSource().getHeaders());
                } else {
                    mediaPlayer.setDataSource(context, getCurrentDataSource().getUri());
                }
            } else if (getCurrentDataSource().getIMediaDataSource() != null) {
                mediaPlayer.setDataSource(getCurrentDataSource().getIMediaDataSource());
            } else if (getCurrentDataSource().getFileDescriptor() != null) {
                mediaPlayer.setDataSource(getCurrentDataSource().getFileDescriptor());
            } else {
                Toast.makeText(context, context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show();
                onError(mediaPlayer, -2, -2);
                return;
            }
            mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
            mediaPlayer.setScreenOnWhilePlaying(true);
            mediaPlayer.prepareAsync();
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(context, context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show();
            onError(mediaPlayer, -2, -2);
        }
    }

    @Override
    public void pause() {
        if (mediaPlayer != null) {
            mediaPlayer.pause();
        }
    }

    @Override
    public boolean isPlaying() {
        if (mediaPlayer == null) {
            return false;
        }
        if (mediaPlayer != null) {
            return mediaPlayer.isPlaying();
        }
        return false;
    }

    @Override
    public void seekTo(long time) {
        if (mediaPlayer == null) {
            return;
        }
        mediaPlayer.seekTo((int) time);
    }

    @Override
    public void release() {
        if (mediaPlayer != null) {
            mediaPlayer.release();
            mediaPlayer = null;
        }
    }

    @Override
    public long getCurrentPosition() {
        if (mediaPlayer == null) {
            return 0;
        }
        return mediaPlayer.getCurrentPosition();
    }

    @Override
    public int getBufferedPercentage() {
        if (mediaPlayer == null) {
            return 0;
        }
        return mediaPlayer.getBufferedPercentage();
    }

    @Override
    public long getDuration() {
        if (mediaPlayer == null) {
            return 0;
        }
        return mediaPlayer.getDuration();
    }

    @Override
    public void setSurface(Surface surface) {
        if (mediaPlayer == null) {
            return;
        }
        mediaPlayer.setSurface(surface);
    }

    @Override
    public void onPrepared(IMediaPlayer mediaPlayer) {
        if (mediaPlayer == null) {
            return;
        }
        if (!isPreparedPause) {
            mediaPlayer.start();
            if (easyMediaManager.getMediaPlay() == this)
                easyMediaManager.getHandlePlayEvent().onPrepared();
        } else {
            isPreparedPause = false;
        }
    }

    @Override
    public void onCompletion(IMediaPlayer mediaPlayer) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().onCompletion(this);
    }

    @Override
    public void onBufferingUpdate(IMediaPlayer mediaPlayer, final int percent) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().setBufferProgress(percent);
    }

    @Override
    public boolean onError(IMediaPlayer mediaPlayer, final int what, final int extra) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().onError(what, extra);
        return true;
    }

    @Override
    public boolean onInfo(IMediaPlayer mediaPlayer, final int what, final int extra) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().onInfo(what, extra);
        return false;
    }

    @Override
    public void onVideoSizeChanged(IMediaPlayer mediaPlayer, int width, int height, int sar_num, int sar_den) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().onVideoSizeChanged(width, height);
    }

    @Override
    public void onSeekComplete(IMediaPlayer mp) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().onSeekComplete();
    }
}

File: ExoPlay2/src/main/java/com/ashlikun/media/exoplay2/play/EasyVideoExo2.kt
package com.ashlikun.media.exoplay2.play

import android.media.AudioManager
import android.net.Uri
import android.view.Surface
import android.widget.Toast
import com.ashlikun.media.R
import com.ashlikun.media.exoplay2.IjkExo2MediaPlayer
import com.ashlikun.media.video.EasyMediaInterface
import com.ashlikun.media.video.EasyMediaManager
import com.ashlikun.media.video.VideoUtils
import tv.danmaku.ijk.media.player.IMediaPlayer
import java.io.File

/**
 * @author　　: 李坤
 * 创建时间: 2023/6/1 17:39
 * 邮箱　　：496546144@qq.com
 *
 * 功能介绍：实现Exo3的播放引擎
 */
class EasyVideoExo2(override val manager: EasyMediaManager) : EasyMediaInterface(manager), IMediaPlayer.OnPreparedListener,
    IMediaPlayer.OnCompletionListener,
    IMediaPlayer.OnBufferingUpdateListener, IMediaPlayer.OnErrorListener, IMediaPlayer.OnInfoListener, IMediaPlayer.OnSeekCompleteListener,
    IMediaPlayer.OnVideoSizeChangedListener {
    /**
     * 是否调用过暂定，调用后在准备好的时候不能直接播放
     */
    private var isPreparedPause = false
    var mediaPlayer: IjkExo2MediaPlayer? = null
    override fun setPreparedPause(isPreparedPause: Boolean) {
        this.isPreparedPause = isPreparedPause
    }

    override fun start() {
        //暂停不在同一个管理器的播放器
        EasyMediaManager.pauseOther(manager)
        mediaPlayer?.start()
    }

    override fun stop() {
        mediaPlayer?.stop()
    }

    override fun prepare() {
        runCatching {
            if (isPlaying) stop()
            release()
        }
        mediaPlayer = IjkExo2MediaPlayer(VideoUtils.context)
        if (currentDataSource == null) {
            Toast.makeText(VideoUtils.context, VideoUtils.context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show()
            onError(mediaPlayer!!, -2, -2)
            return
        }
        mediaPlayer!!.setAudioStreamType(AudioManager.STREAM_MUSIC)
        mediaPlayer!!.setOnPreparedListener(this@EasyVideoExo2)
        mediaPlayer!!.setOnCompletionListener(this@EasyVideoExo2)
        mediaPlayer!!.setOnBufferingUpdateListener(this@EasyVideoExo2)
        mediaPlayer!!.setScreenOnWhilePlaying(true)
        mediaPlayer!!.setOnErrorListener(this@EasyVideoExo2)
        mediaPlayer!!.setOnInfoListener(this@EasyVideoExo2)
        mediaPlayer!!.setOnVideoSizeChangedListener(this@EasyVideoExo2)
        mediaPlayer!!.setOnSeekCompleteListener(this)
        mediaPlayer!!.isLooping = currentDataSource!!.isLooping
        //通过自己的内部缓存机制
        mediaPlayer!!.isCache = currentDataSource!!.isCache
        if (currentDataSource!!.cacheDir.isNotEmpty()) {
            mediaPlayer!!.cacheDir = File(currentDataSource!!.cacheDir)
        }
        mediaPlayer!!.overrideExtension = currentDataSource!!.overrideExtension
        try {
            if (currentDataSource!!.url.isNotEmpty()) {
                if (currentDataSource!!.headers != null) {
                    mediaPlayer!!.setDataSource(VideoUtils.context, Uri.parse(currentDataSource!!.url), currentDataSource!!.headers)
                } else {
                    mediaPlayer!!.dataSource = currentDataSource!!.url
                }
            } else if (currentDataSource!!.uri != null && currentDataSource!!.uri.toString().isNotEmpty()) {
                if (currentDataSource!!.headers != null) {
                    mediaPlayer!!.setDataSource(VideoUtils.context, currentDataSource!!.uri, currentDataSource!!.headers)
                } else {
                    mediaPlayer!!.setDataSource(VideoUtils.context, currentDataSource!!.uri)
                }
            } else if (currentDataSource!!.iMediaDataSource != null) {
                mediaPlayer!!.setDataSource(currentDataSource!!.iMediaDataSource)
            } else if (currentDataSource!!.fileDescriptor != null) {
                mediaPlayer!!.setDataSource(currentDataSource!!.fileDescriptor)
            } else {
                Toast.makeText(VideoUtils.context, VideoUtils.context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show()
                onError(mediaPlayer!!, -2, -2)
                return
            }
            mediaPlayer!!.setAudioStreamType(AudioManager.STREAM_MUSIC)
            mediaPlayer!!.setScreenOnWhilePlaying(true)
            mediaPlayer!!.prepareAsync()
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(VideoUtils.context, VideoUtils.context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show()
            onError(mediaPlayer!!, -2, -2)
        }
    }

    override fun pause() {
        mediaPlayer?.pause()
    }

    override val isPlaying: Boolean
        get() = mediaPlayer?.isPlaying ?: false

    override fun seekTo(time: Long) {
        mediaPlayer?.seekTo(time.toInt().toLong())
    }

    override fun release() {
        mediaPlayer?.release()
        mediaPlayer = null
    }

    override val currentPosition: Long
        get() = mediaPlayer?.currentPosition ?: 0

    override val bufferedPercentage: Int
        get() = mediaPlayer?.bufferedPercentage ?: 0

    override val duration: Long
        get() = mediaPlayer?.duration ?: 0

    override fun setSurface(surface: Surface) {
        mediaPlayer?.setSurface(surface)
    }

    override fun onPrepared(mediaPlayer: IMediaPlayer) {
        if (!isPreparedPause) {
            mediaPlayer.start()
            if (manager.mediaPlay === this) manager.handlePlayEvent.onPrepared()
        } else {
            isPreparedPause = false
        }
    }

    override fun onCompletion(mediaPlayer: IMediaPlayer) {
        if (manager.mediaPlay === this) manager.handlePlayEvent.onCompletion(this)
    }

    override fun onBufferingUpdate(mediaPlayer: IMediaPlayer, percent: Int) {
        if (manager.mediaPlay === this) manager.handlePlayEvent.setBufferProgress(percent)
    }

    override fun onError(mediaPlayer: IMediaPlayer, what: Int, extra: Int): Boolean {
        if (manager.mediaPlay === this) manager.handlePlayEvent.onError(what, extra)
        return true
    }

    override fun onInfo(mediaPlayer: IMediaPlayer, what: Int, extra: Int): Boolean {
        if (manager.mediaPlay === this) manager.handlePlayEvent.onInfo(what, extra)
        return false
    }

    override fun onVideoSizeChanged(mediaPlayer: IMediaPlayer, width: Int, height: Int, sar_num: Int, sar_den: Int) {
        if (manager.mediaPlay === this) manager.handlePlayEvent.onVideoSizeChanged(width, height)
    }

    override fun onSeekComplete(mp: IMediaPlayer) {
        if (manager.mediaPlay === this) manager.handlePlayEvent.onSeekComplete()
    }
}